cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerService.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.worker.system;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import io.cdap.cdap.api.metrics.MetricsCollectionService;
+import io.cdap.cdap.api.service.worker.RunnableTask;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.SConfiguration;
+import io.cdap.cdap.common.discovery.ResolvingDiscoverable;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.http.CommonNettyHttpServiceBuilder;
+import io.cdap.cdap.common.security.HttpsEnabler;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactManagerFactory;
+import io.cdap.cdap.internal.app.worker.RunnableTaskLauncher;
+import io.cdap.http.ChannelPipelineModifier;
+import io.cdap.http.NettyHttpService;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.http.HttpContentDecompressor;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.DiscoveryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Launches an HTTP server for receiving and handling {@link RunnableTask}
+ */
+public class SystemWorkerService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemWorkerService.class);
+  private static final Gson GSON = new Gson();
+
+  private final CConfiguration cConf;
+  private final DiscoveryService discoveryService;
+  private final NettyHttpService httpService;
+  private final ArtifactManagerFactory artifactManagerFactory;
+  private final RunnableTaskLauncher taskLauncher;
+  private Cancellable cancelDiscovery;
+  private InetSocketAddress bindAddress;
+  private MetricsCollectionService metricsCollectionService;
+
+  @Inject
+  SystemWorkerService(CConfiguration cConf,
```

Comment:
same question here. What's the delta between this and existing service? Can they share the same code? Or can we extract the common code to a base class?

Reply:
This one should be possible. I can check this and get back. 

One important reason why we didnt refactor code to move to common interface was that we don't want the flows to be interconnected. Any changes to worker pod flow should not affect system pods and vice versa. We think they will evolve separately over a period of time.

Reply:
Understand. We can always split and diverge when needed. At the moment, there is a high degree of duplication. Let's see if we can extract common or reuse existing code.

Reply:
Discussed over call. @suyashahuja02 to follow up with a refactoring PR